### DISCUSSION TITLE
Flash Lanbon L8 without opening

### DISCUSSION THREAD

**Discussion by nagyrobi** [HASwitchPlate/openHASP#76] [Hardware]
No need to open up the device and no need to solder to the board anything. You need to prepare some thin wires as standard dupont pins don't fit in the connector. For example this is made by soldering some waste 0.5W resistor feet to dupont wires:
![kép](https://user-images.githubusercontent.com/1550668/110013874-6f06aa80-7d22-11eb-9162-b41f409bccb6.png)

First insert a free wire in IO0 pin (leftmost in the lower row):
![kép](https://user-images.githubusercontent.com/1550668/110014160-c442bc00-7d22-11eb-9ef7-cb987b5f5f62.png)

Then insert RX, TX, 5V and GND in the upper row, see the markings on the pcb. You can also use 3.3V flash, but I think 5V is better if your USB-TTL adapter offers this because the display also turns on so after flashing you can see immediately if it was successful or not, and you can connect to serial port with a terminal client to see debug messages if you want to. Watch out not to short-circuit any of them:
![kép](https://user-images.githubusercontent.com/1550668/110014451-184da080-7d23-11eb-9479-5e1616059c77.png)

While you plug in your USB-TTL adapter keep the free cable coming from IO0 to GND, you can release afterwards.
![kép](https://user-images.githubusercontent.com/1550668/110014556-3915f600-7d23-11eb-8211-2adf79cd5196.png)

Erase Flash as usual (commands below for Ubuntu):
`./esptool.py --port /dev/ttyUSB3 erase_flash`

Disconnect the USB-TTL, reconnect again with IO0 to GND like before and write Flash:
`./esptool.py --port /dev/ttyUSB3 --baud 460800 write_flash 0x0 lanbon_l8_full_8MB_<version>.bin`

Disconnect the USB-TTL, reconnect, and see what's happening. Watch the bootup procedure with `screen /dev/ttyUSB3 115200`. 

Good luck!

**Comment by fvanroie** [comment]
Can we add these pictures and information to https://fvanroie.github.io/hasp-docs/#devices/lanbon-l8 ?

**Reply by nagyrobi** [reply]
Of course!

**Reply by fvanroie** [reply]
Question: Would it be possible to make some 800x480-ish photo's like the last one?
Resizing and cropping these to fit wouldn't scale as nicely.

**Reply by nagyrobi** [reply]
I think simply scaling them to the same width is enough, height is content-dependent... I specifically cropped down parts from top/bottom of these because they were just non-informative white plastic surfaces.

**Comment by erwin** [comment]
Tips from the perspective of a software guy with very limited hardware experience trying to get this working.

You'll need to "connect".

`RX`, `TX`, `5V`, `GND` (ground) and `GPIO0`

## Important: Wire Size

Getting wires that are the correct size for the lanbon's ports can be a bit tricky. I bought a few different wire sizes they were incorrect. Finally on my 3rd or 4th order I managed to get pins that were the right size.

```text
| USB TTL | → | Lanbon L8 |   |
|:-------:|---|:---------:|---|
|   GND   | → |    GND    | ╮ |
|    5V   | → |     5V    | │ |
|    TX   | → |     RX    | │ |
|    RX   | → |     TX    | │ |
|         |   |    IO0    | ╯ |

```

To get a cable to connect the Lanbon's `IO0` to the Lanbon's `GND` I sliced open a Cat-6 ethernet cable and used the individual stands from inside that. 

If you're wires are too big you'll damage the connectors (which are also required for daily operation!) and if your wires are too small, they'll just fall out and you'll have an immensely difficult time flashing.

Note that when flashing, these voltages are very low. When my fingertips were touching the `GND` pin as I was holding `IO0` onto the `GND` pin, it never shocked me. 

# Example "Wiring Harness" photos:

![image](https://user-images.githubusercontent.com/11534217/155832919-5520dc07-4e3c-466b-b88d-27cbc00cfbd0.png)
![image](https://user-images.githubusercontent.com/11534217/155832939-b75de6b9-c3b3-48ca-9ff9-30bd7e38de93.png)
![image](https://user-images.githubusercontent.com/11534217/155832966-190f1108-f1c3-4df1-b459-5ba13a5bc420.png)

# Connecting and Flashing Step-by-Step

1. It's a little easier to connect the wires with the back cover out of the way. If you have trouble getting the pins in, just remove the 4 screws and remove the back cover. Be gentle though, there is a wiring connecting the two sides of the plate that you don't want to tear out.
2. Disconnect `USB TTL` from your Computer's USB Port
3. Connect a loose wire to `IO0`, this is on the bottom row of pins so we'll connect it first, and we must manually connect it to GROUND while we power cycle... Just leave the other end hanging for now.
4. Connect `GND` on USB TTL to `GND` on Lanbon
5. Connect `5V` on USB TTL to `5V` on Lanbon
6. Connect `TXD` on USB TTL to `RX` on Lanbon
7. Connect `RXD` on USB TTL to `TX` on Lanbon
8. Double check that the `IO0` wire still has one end inside of the `IO0` port.
9. Loosen the `GND` wire a little (pull it out about 2mm) and connect the other end of the `IO0` wire to the `GND` wire. DO NOT LET GO... Hold these.
10. Now, boot the Lanbon into Flashing Mode by connecting the `USB TTL` into Computer's USB Port
11. Once the Lanbon boots to a solid WHITE SCREEN, that's the indicator that it's in **flashing mode**. If it boots anything else, such as the default software, double check each step, especially:
    - Step 9 Lanbon `IO0` connected to Lanbon `GND`
    - Step 6 `USB TTL TXD` connected to Lanbon `TX`
    - Step 7 `USB TTL RXD` conneted to `Lanbon TX`.
13. Stop holding the `IO0` wire to the `GND` wire.
14. While everything else is still connected, run `sudo esptool.py` or whatever the flashing tool on your platform is. For me the command was:

```
sudo esptool.py --port /dev/ttyUSB0 --baud 460800 write_flash 0x0 ~/Downloads/lanbon_l8_full_8MB_v0.6.2.bin
```

`esptool` output when successful looks something like:

```
esptool.py v3.2
Serial port /dev/ttyUSB0
Connecting......
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting...
Detecting chip type... ESP32
Chip is ESP32-D0WD (revision 1)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 34:86:5d:42:b6:9c
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Flash will be erased from 0x00000000 to 0x001abfff...
Compressed 1749632 bytes to 993268...
Wrote 1749632 bytes (993268 compressed) at 0x00000000 in 24.0 seconds (effective 582.3 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

The important thing is that you'll see data being successfully transferred.

1. If you're not getting the `white screen`, double check each wire connection
2. If `esptool` isn't flashing the Lanbon, remove and replug the USB-TTL (make sure the Lanbon still comes up with the White Screen (flashing mode).
3. If you flashed it, but it still doesn't boot, make sure you're using the correct firmware: `lanbon_l8_full`-something...

I flashed 5 more of these today using the process above and it worked great for me.

**Comment by janick** [comment]
Has anyone flashed a L8 with the new [0.7.0 firmware](https://www.openhasp.com/0.7.0/firmware/) yet?

And if so, which of the three binaries did you use?

**Reply by erwin** [reply]
* Currently all my Lanbon L8's are on `OpenHASP 0.7.0-rc4`
* Looking through the commits `rc12` was 2 months ago
* I don't see an official `0.7` release tag...

That said, mine on `0.7.0-rc4` are stable enough most of the time.

**Reply by janick** [reply]
Thank you. And did you use one of the pre-built binaries (in which case which one?) or did you build your own (in which case would you mind adding it to the repo)?

**Reply by erwin** [reply]
I build my own but I include the a two Chinese fonts in the bundle that others may not want.
When I next rebuild though, I could try to do a stock rebuild as well and share that image.

Here's the current image that I'm running, but when I re-build I will post new ones...
I can't remember if it's the `full` or the `ota` version that works with the lanbon when doing the update via the Lanbon's OpenHASP WebUI.


[lanbon_l8_.zip](https://github.com/user-attachments/files/15991024/lanbon_l8_.zip)

**Reply by fvanroie** [reply]
Automated builds can be downloaded from [Github Actions](https://github.com/HASwitchPlate/openHASP/actions) or either https://nightly.openhasp.com

**Reply by janick** [reply]
> Automated builds can be downloaded from [Github Actions](https://github.com/HASwitchPlate/openHASP/actions) or either https://nightly.openhasp.com

Awesome! Is there a link to that nightly build page somewhere on the http://openhasp.com site? I looked everywhere and I know I didn't find it.

**Reply by fvanroie** [reply]
e.g. [Flash ESP32](https://www.openhasp.com/0.7.0/firmware/esp32/?h=nightly#online-installation)
If you have suggestions for improving the documentation, you can click the pen icon at the top of the page.

**Comment by nagyv** [comment]
Did anyone run into issues with flashing the L8-HS lately?

After starting in debug mode:

```
$ esptool.py --port /dev/cu.usbserial-0001 erase_flash
esptool.py v4.8.1
Serial port /dev/cu.usbserial-0001
Connecting......................................

A fatal error occurred: Failed to connect to Espressif device: No serial data received.
For troubleshooting steps visit: https://docs.espressif.com/projects/esptool/en/latest/troubleshooting.html
```

When starting in "normal" mode, minicom can connect:

```
Welcome to minicom 2.10

OPTIONS: I18n 
Compiled on Jan  1 1980, 00:00:00.
Port /dev/tty.usbserial-0001, 15:20:30 [F]

Press CTRL-A Z for help on special keys

I (2339) spiram: SPI SRAM memory test OK
I (2341) heap_init: Initializing. RAM available for dynamic allocation:
I (2341) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (2346) heap_init: At 3FFC42C8 len 0001BD38 (111 KiB): DRAM
I (2352) heap_init: At 3FFE0440 len 0001FBC0 (126 KiB): D/IRAM
I (2359) heap_init: At 40078000 len 00008000 (32 KiB): IRAM
I (2365) heap_init: At 4009F2D8 len 00000D28 (3 KiB): IRAM
I (2371) cpu_start: Pro cpu start user code
I (2376) spiram: Adding pool of 4077K of external SPI memory to heap allocator
I (48) esp_core_dump_flash: Init core dump to flash
I (58) esp_core_dump_flash: Found partition 'coredump' @ bc6000 65536 bytes
E (67) esp_core_dump_common: Failed to mmap core dump data (260)!
I (68) cpu_start: Starting scheduler on PRO CPU.
I (70) spiram: Reserving pool of 32K of internal memory for DMA/internal allocations
W (91) lanbon_L8 - dataManage: nvs_data devSysKeyParam read success, dataTemp_devSysKeyParam.sysParam_tuyamoudleExsistFlg=0

>>>sed objChg_bitHold:00.
>>>sed objChg_bitHold:00.
temprature caliVal set val:0.000000.
W (110) lanbon_L8 - dataManage: nvsItem:'system_selected' read:0.
W (111) lanbon_L8 - dataManage: nvsItem:'system_running' read res:0, len:8
W (115) lanbon_L8 - dataManage: nvsItem:'system_running' read:nothing.
W (123) lanbon_L8 - dataManage: nvsItem:'part_idx_slot' read:{00, 01, 02}.
W (130) lanbon_L8 - dataManage: partition run current == NULL.
W (136) lanbon_L8 - dataManage: deviceParamPrtBoot->selAvailable:true
>>>>>>>>>>pcnt isr reg res:0.
eFuse Two Point: NOT supported
eFuse Vref: Supported
Characterized using eFuse Vref
I (154) gpio: GPIO[12]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (163) gpio: GPIO[14]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (172) gpio: GPIO[27]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (387) gpio: GPIO[22]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
>>>Display LCD ID:00A94200.
W (6410) lanbon_L8 - uiSysPreCfg: localRunningParam.paramBootPrt.selAvailable:true
```

Connecting in "debug" mode works too, but  ... well there is nothing to copy. :)
